feat(docs): explanation page about policy groups. (#453)
kubewarden/docs: docs/explanations/policy-groups.md

---
sidebar_label: Policy Groups
sidebar_position: 21
title: Policy Groups
description: A description of Kubewarden policy groups
keywords: [ kubewarden, policy groups, clusteradmissionpolicygroup, admissionpolicygroup]
doc-persona: [kubewarden-operator]
doc-type: [explanation]
doc-topic: [explanations, policy-group]
---

<head>
  <link rel="canonical" href="https://docs.kubewarden.io/explanations/policy-groups"/>
</head>

The policy group feature allows users to create complex policies by combining
multiple simpler ones. It introduces two new Custom Resource Definitions
(CRDs):

- `AdmissionPolicyGroup`: For admission policies that apply to specific
  namespaces. 
- `ClusterAdmissionPolicyGroup`: For admission policies that apply across the
  entire cluster.

> These policy groups enable users to leverage existing policies, reducing the

Reviewer: ```suggestion
These policy groups enable users to use existing policies, reducing the
```

need for custom policy creation and enhancing reusability. By avoiding
duplication of policy logic, users can simplify management and create custom
policies with a DSL-like configuration.

> The primary goal of policy groups is to facilitate the composite evaluation of
> multiple policies using a boolean expression, allowing for the definition of
> complex logic. However, it is important to note that while ordinary policies
> can include mutation logic to modify resources during admission, policy groups
> are limited to validation only.

Reviewer: ```suggestion
Policy groups enable the combined evaluation of
multiple policies using a boolean expression. This allows the definition of
complex logic. However, it is important to note that while ordinary policies
can include mutation logic to modify resources during admission, policy groups
are limited to validation only.
```

> Most of the configuration for policy groups is similar to that of ordinary
> policies. The main differences include the addition of the `expression`,
> `message`, and `policies` fields, as well as the declaration of context-aware
> rules in a different location.

Reviewer: ```suggestion
Configuration for policy groups is similar to that of ordinary
policies. The difference is the addition of the `expression`,
`message`, and `policies` fields, as well as the declaration of context-aware
rules in a different location.
```

## `policies` in a Policy Group

> The policies field is a map of ordinary policies (`ClusterAdmissionPolicy` and
> `AdmissionPolicy`) that make up the policy group. These are other Kubewarden
> policies invoked by the policy group to determine whether the resource under
> evaluation is accepted or rejected. The definitions of these policies are a
> simplified version of the ordinary Kubewarden policies, containing only the
> `module`, `settings`, `contextAwareResources`, and `name` of the policy. These
> elements are necessary for the policies to function within a policy group,
> while all other configurations are defined within the policy group itself, as
>     the policy group is considered a single policy. 

Reviewer: ```suggestion
The policies field is a map of ordinary policies (`ClusterAdmissionPolicy` and
`AdmissionPolicy`) that make up the policy group. Kubewarden
policies are called by the policy group, to determine whether the resource under
evaluation is accepted or rejected. The definitions of these policies are a
simplified version of ordinary Kubewarden policies, containing only the
`module`, `settings`, `contextAwareResources`, and `name` of the policy. These
elements are necessary for the policies to function within a policy group.

```

I've taken this bit out 

'while all other configurations are defined within the policy group itself, as
the policy group is considered a single policy.' 

I'm not sure what you are trying to say here. Maybe a little rewording is necessary.?

## Policy Group `expression` and `message`

> Policy groups allow users to perform complex evaluations using the policies
> defined in the policies field by utilizing

Reviewer: ```suggestion
Policy groups, in the policies field, allow users to perform complex evaluations using
```

> [CEL](https://github.com/google/cel-go) as the expression language. This choice
> is made because CEL is also used by the `ValidatingAdmissionPolicy` and
> `matchConditions` in Kubernetes. Here is an example of a

Reviewer: ```suggestion
[CEL](https://github.com/google/cel-go) as the expression language. CEL is used
because CEL is used by Kubernetes in the `ValidatingAdmissionPolicy` and
`matchConditions`. Here is an example of a
```

        - UPDATE
  backgroundAudit: true
  policies:
    sigstore_pgp:
      module: ghcr.io/kubewarden/policies/verify-image-signatures:v0.2.8
      settings:
        signatures:
          - image: "*"
            pubKeys:
              - "-----BEGIN PUBLIC KEY-----xxxxx-----END PUBLIC KEY-----"
              - "-----BEGIN PUBLIC KEY-----xxxxx-----END PUBLIC KEY-----"
    sigstore_gh_action:
      module: ghcr.io/kubewarden/policies/verify-image-signatures:v0.2.8
      settings:
        signatures:
          - image: "*"
            githubActions:
              owner: "kubewarden"
    reject_latest_tag:
      module: ghcr.io/kubewarden/policies/trusted-repos-policy:v0.1.12
      settings:
        tags:
          reject:
            - latest
  expression: "sigstore_pgp() || (sigstore_gh_action() && reject_latest_tag())"
  message: "The policy group is rejected."
```

In this example, each policy in the group is represented as a function call in
the `expression`, using the same name as the policy defined in the group. The
`expression` field must be a valid CEL `expression` that evaluates to a boolean
value, and it will be validated by the Kubewarden controller's webhook. If the

> `expression` evaluates to true, the group policy is considered accepted;

Reviewer: ```suggestion
`expression` evaluates to true, the group policy is accepted;
```

otherwise, it is rejected.

The message field specifies the message returned when the policy group is
rejected. The specific results of the policies will be included in the warning
field of the response.

Here is an example of the response returned when the policy group is rejected:

```json
{
  "apiVersion": "admission.k8s.io/v1",
  "kind": "AdmissionReview",
  "response": {
    "uid": "<value from request.uid>",
    "allowed": false,
    "warning": [
      "sigstore_pgp was rejected: the image signature is not valid",
      "sigstore_gh_action was accepted",
      "latest_tag was rejected: the image tag latest is not allowed"
    ]
  }
}
```

Currently, there is no production-ready CEL library available for Rust. To work
around this limitation and emulate CEL's behavior and syntax, Kubewarden uses
[Rhai](https://rhai.rs/), a small, fast, and easy-to-use scripting language and

> evaluation engine that integrates seamlessly with Rust. Rhai is easy to embed
> in policy server, allows for the creation of a engine with a bare minimum set
> of functionalities, does not introduce significant overhead. 

Reviewer: ```suggestion
evaluation engine that integrates seamlessly with Rust. Rhai is easy to embed
in the Kubewarden policy server, allows for the creation of an engine with a minimum set
of functionalities, and doesn't introduce significant overhead. 
```

Considering the expression from the example above:

```cel 
sigstore_pgp() || (sigstore_gh_action() && reject_latest_tag())
```

Each policy in the group will be represented as a function call in the
expression with the same name as the policy defined in the group.

> The functions call are executed only if necessary. In other words, if the result
> of the first operator in the logic is enough to determine the result of the
> expression, the other operators will not be evaluated. This approach avoids
> unnecessary evaluations of all policies in the group when the result is already
> known.

Reviewer: ```suggestion
The functions are called only if necessary. If the result
of the first operator in the logic is enough to determine the result of the
expression, the other operators will not be evaluated. This avoids
unnecessary evaluations of policies in the group. 
```

## Context-Aware Rules

Another distinction between policy groups and ordinary policies is the location
where context-aware resource rules are defined. Each policy in a group can
accept an optional `contextAwareResources` field to specify the resources that
the policy is allowed to access during evaluation.

**Example:**
```yaml
apiVersion: policies.kubewarden.io/v1
kind: ClusterAdmissionPolicyGroup # or AdmissionPolicyGroup
metadata:
  name: context-aware-group
spec:
  rules:
    - apiGroups: [""]
      apiVersions: ["v1"]
      resources: ["pods"]
      operations:
        - CREATE
        - UPDATE
  policies:
    policy_1:
      module: ghcr.io/kubewarden/policies/policy_1:v0.1.0
      contextAwareResources:
        - apiVersion: "v1"
          kind: "Pod"
      settings:
        foo: "bar"
    policy_2:
      module: ghcr.io/kubewarden/policies/policy_2:v0.1.0
      contextAwareResources:
        - apiVersion: "v1"
          kind: "Namespace"
      settings:
        foo: "bar"
  expression: "policy_1() && policy_2()"
  message: "The policy group is rejected."
```

## Settings Validation

When the policy server starts, it will validate the settings of both policy

> groups and ordinary policies. However, policy groups will undergo an additional

Reviewer: ```suggestion
groups and ordinary policies. However, policy groups undergo an additional
```

validation step to ensure that the expression is valid and evaluates to a
boolean value.

## Audit Scanner

> Similar to the AdmissionPolicy and ClusterAdmissionPolicy CRDs, the
> backgroundAudit field will indicate whether the policy group should be included
> or excluded during audit checks.

Reviewer: ```suggestion
Similar to the AdmissionPolicy and ClusterAdmissionPolicy CRDs, the
backgroundAudit field indicates if the policy group should be included
during audit checks.
```

## Policy Server

> The policies.yaml settings file will be extended to include policy groups
> alongside ordinary policies. As with ordinary policies, modules will be
> downloaded once. Therefore, if the same policy module is used in both a policy
> group and an ordinary policy, the same module will be utilized.

Reviewer: ```suggestion
The `policies.yml` settings file is extended to include policy groups
alongside ordinary policies. As with ordinary policies, modules are
downloaded once. The same policy module is used in both a policy
group and an ordinary policy.
```

I may be assuming wrongly about the file extension in use.

---
sidebar_label: Policy Groups
sidebar_position: 21
title: Policy Groups
description: A description of Kubewarden policy groups
keywords: [kubewarden, policy groups, clusteradmissionpolicygroup, admissionpolicygroup]
doc-persona: [kubewarden-operator]
doc-type: [explanation]
doc-topic: [explanations, policy-group]
---

<head>
  <link rel="canonical" href="https://docs.kubewarden.io/explanations/policy-groups"/>
</head>

The policy group feature allows users to create complex policies by combining

> multiple simpler ones. It introduces two new Custom Resource Definitions

Reviewer 2: maybe `multiple` is redundant

evaluation is accepted or rejected. The definitions of these policies are a
simplified version of ordinary Kubewarden policies, containing only the
`module`, `settings` and `contextAwareResources` attributes. These
elements are necessary for the policies to function within a policy group.

Each policy of the group policy is identified by a unique name. For example,
the following snippet defines two policies: `signed_by_alice`, `signed_by_bob` and `reject_latest_tag`.

```yaml
policies:
  signed_by_alice:
    module: ghcr.io/kubewarden/policies/verify-image-signatures:v0.2.8
    settings: {} # settings for the policy
  signed_by_bob:
    module: ghcr.io/kubewarden/policies/verify-image-signatures:v0.2.8
    settings: {} # settings for the policy
  reject_latest_tag:
    module: ghcr.io/kubewarden/policies/trusted-repos-policy:v0.1.12
    settings: {} # settings for the policy
```

:::tip
The same policy can be included multiple times in the same policy group, with
different settings.
:::

### The `expression` attribute

The `expression` attribute contains a statement made of the policy
identifiers joined together by logical operators.

> The evaluation of the `expression` statement must produce a boolean value.

Reviewer 2: `must evaluate to a boolean value`

inside of the `.spec.policies` map. The results produced
by the evaluation of the policies are then evaluated using the logical operators
provided by the user.

These are the supported operators:

- `&&`: used to perform `AND` operations
- `||`: used to perform `OR` operations
- `!`: used to perform `NOT` operations

Round brackets `( )` can be used to define evaluation priorities.

For example, given the following expression:

```yaml
reject_latest() || (signed_by_alice() && signed_by_bob())
```

The policy will reject workloads that have images using the `latest` tag, unless
these images are signed both by Alice and Bob.

### The `message` attribute and the response format

The `message` field specifies the message returned when the evaluation of the
`expression` results in a rejection. The message is included in the response,
together with the results of the individual policies evaluation.

Group Policies rely on the `warnings` attribute of the
[`AdmissionReview` response](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#response)
object to provide information about the evaluation results of the
policies that are part of the group.

> These warnings are shown by multiple kubernetes clients, including `kubectl`.

Reviewer: ```suggestion
These warnings are shown by multiple Kubernetes clients, including `kubectl`.
```

For example, this is the output produced when attempting to create a Pod with an image that

> uses the `latest` and is signed only by Alice:

Reviewer: ```suggestion
uses the `latest` tag and is signed only by Alice:
```

        "field": "spec.policies.signed_by_bob"
      },
      {
        "message": "not allowed, reported errors: tags not allowed: latest",
        "field": "spec.policies.reject_latest"
      }
    ]
  },
  "code": 400
}]
Error from server: error when creating "signed-pod.yml": admission webhook "clusterwide-demo.kubewarden.admission" denied the request: the image is using the latest tag or is not signed by Alice and Bob
```

The full admission response is available in the logs of the Policy Server
when running in debug mode.
Moreover, the evaluation details are always part of the OpenTelemetry traces emitted by Policy Server.

## Context-Aware Policies

Another distinction between policy groups and ordinary policies is the location

> where context-aware resource rules are defined. Each policy in a group can
> accept an optional `contextAwareResources` field to specify the resources that

Reviewer: ```suggestion
where context-aware resource rules are defined. Each policy in a group
accepts an optional `contextAwareResources` field to specify the resources that
```
